pandas/core/groupby/generic.py

```diff
@@ -1616,20 +1507,27 @@ def count(self):
         DataFrame
             Count of values within each group.
         """
-        data, _ = self._get_data_to_aggregate()
+        output = OrderedDict()
+        names = []
+
+        # TODO: dispatch to _cython_agg_general instead of custom looping
+        # TODO: refactor with series logic
         ids, _, ngroups = self.grouper.group_info
-        mask = ids != -1
 
-        val = (
-            (mask & ~_isna_ndarraylike(np.atleast_2d(blk.get_values())))
-            for blk in data.blocks
-        )
-        loc = (blk.mgr_locs for blk in data.blocks)
+        if self.axis == 0:
+            iter_obj = self._obj_with_exclusions
+        else:
+            iter_obj = self._obj_with_exclusions.T
 
-        counter = partial(lib.count_level_2d, labels=ids, max_bin=ngroups, axis=1)
-        blk = map(make_block, map(counter, val), loc)
+        for index, (name, obj) in enumerate(iter_obj.items()):
```

Comment:
This isn't the ideal way to do this, but to keep the scope of the PR minimum and get tests passing I put this iteration directly in this method. This really should be shared by a few aggregation functions, but I believe is currently blocked by #25610 and #21668, amongst possibly others

Would prefer to address comprehensively in a follow up